arangod/Scheduler/SupervisedScheduler.cpp

```diff
@@ -211,25 +223,35 @@ bool SupervisedScheduler::queue(RequestLane lane, std::function<void()> handler,
   // use memory order release to make sure, pushed item is visible
   uint64_t jobsSubmitted = _jobsSubmitted.fetch_add(1, std::memory_order_release);
   uint64_t approxQueueLength = jobsSubmitted - _jobsDone;
+  // fetching _jobsDone after _jobsSubmitted means there can be data races. 
+  // if the current thread doesn't get interrupted, the condition 
+  //   _jobsSubmitted > _jobsDone   will always hold true
+  // however, if the current thread gets interrupted and other threads finish their 
+  // jobs first, then the reverse may also be true:   jobsSubmitted < _jobsDone
+  // we guard against this by checking approxQueueLength for an underlow
+  if (ADB_UNLIKELY(approxQueueLength > std::numeric_limits<uint64_t>::max() - _maxFifoSize)) {
+    approxQueueLength = 0;
+  }
+
```

Comment:
shouldn't it be enough to load `_jobsDone` into a local variable before `fetch_add` on `_jobsSubmitted`? And we should add a `+ 1` to `jobsSubmitted` because it is fetch and then add.
https://en.cppreference.com/w/cpp/atomic/atomic/fetch_add
> The value immediately preceding the effects of this function in the modification order of *this.

Reply:
Yes, this would also work.

Reply:
@maierlars which operation should be sequentially consistent?

Reply:
@mpoeter consider this part of the question. At the time writing the above statement I thought one could read the a _newer_ value of `_jobsDone`, but an older value of  `_jobsSubmitted` when having both as relaxed operation. It turned out that I have no idea of how this works together with the `_add` part.

Reply:
@jsteemann @maierlars After going through the whole file I came to the following conclusion:
  * the operations on `_jobsDequeued`, `_jobsSubmitted` and `_jobsDirectExec` can all be relaxed (they are not used to synchronize any satellite data),
  * the only member that is used to synchronize satellite data is `_jobsDone` - therefore the `fetch_add` calls should use `memory_order_release`, and the load operations should use `memory_order_acquire`,
  * the `_jobsDone.load` should always be ordered before the `load` operations of any of the other `_jobsXX` members.

I don't see the necessity for any of the operations to be sequentially consistent.